### PR TITLE
Move Docker to Subdirectory

### PR DESCRIPTION
**Pull request by jason-fox**
Same as for JSON Agent.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ARG NODE_VERSION=8.12.0-slim
+FROM node:${NODE_VERSION}
+ARG GITHUB_ACCOUNT=telefonicaid
+ARG GITHUB_REPOSITORY=iotagent-ul
+ARG DOWNLOAD_TYPE=latest
+
+ENV GITHUB_ACCOUNT=${GITHUB_ACCOUNT}
+ENV GITHUB_REPOSITORY=${GITHUB_REPOSITORY}
+ENV DOWNLOAD_TYPE=${DOWNLOAD_TYPE}
+
```
Comment (fgalan) [inline docker/Dockerfile]: What't the purpose of having a parallel set of ARG and ENV? I understand it for some good reason :) but at a glance it a bit redundant and add complexity.

Could you elaborate, please?

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+# 
+# To obtain the latest stable release run this Docker file with the parameters
+# --no-cache --build-arg DOWNLOAD_TYPE=stable
+#
+# Alternatively for local development, just copy this Dockerfile into file the
+# root of the repository and copy over your local source using : 
+#
+# COPY . /opt/iotaul/
+#
+RUN if [ ${DOWNLOAD_TYPE} = "latest" ] ; then RELEASE="master"; else RELEASE=$(curl -s https://api.github.com/repos/${GITHUB_ACCOUNT}/${GITHUB_REPOSITORY}/releases/latest | grep 'tag_name' | cut -d\" -f4); fi && \
```
Comment (fgalan) [inline docker/Dockerfile]: Could this RUN step be merge with the othe one? I undertand doing so would have two advantages:

1. Only one layer in docker image building
2. To remove some redundant steps such as `apt-get clean`, etc.

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  	if [ ${DOWNLOAD_TYPE} = "latest" ] ; then echo "INFO: Building Latest Development"; else echo "INFO: Building Release: ${RELEASE}"; fi && \
+  	apt-get update && \
+  	apt-get install -y  --no-install-recommends unzip && \
+  	wget --no-check-certificate -O source.zip https://github.com/${GITHUB_ACCOUNT}/${GITHUB_REPOSITORY}/archive/${RELEASE}.zip && \
+	unzip source.zip && \
```
Comment (fgalan) [inline docker/Dockerfile]: For cleanness, source.zip should be removed after use.

File: docker/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# FIWARE IoT Agent for the Ultralight 2.0 protocol
```
Comment (fgalan) [inline docker/README.md]: A link to this README.md should be included in the main README.md (as it is done for instance in Orion repository, check https://github.com/telefonicaid/fiware-orion#advanced-topics)

File: docker/Dockerfile (hunk 1 above)
Comment (jason-fox) [reply docker/Dockerfile]: Yes it's kind of ugly, but it is standard practice to obtain build time environment variables. 

`ARG` is only available on build, so setting it as an `ENV` enables it to stick around when the container is running (hence you can set the value with a `--build-arg` and then read it from the running container (so `docker exec <my-image-name> env` should return information about the GitHub repo the image was built against.

* https://vsupalov.com/docker-build-time-env-values/
* https://vsupalov.com/docker-arg-env-variable-guide/

File: docker/Dockerfile (hunk 2 above)
Comment (jason-fox) [reply docker/Dockerfile]: The second `RUN` statement occurs after the `WORKDIR` has been set, hence it is running from a separate directory. The download is effectively just copying a zip to root, exploding it and moving it to the right place. It is the logical equivalent to the `COPY` statement so I've kept it as a separate layer.

As you know I'm trying to do three incompatible things here, and I think that keeping a separate layer equivalent to the `COPY` statement is the simplest way to  be able to maintain a single `Dockerfile` for Automated Builds which can be painlessly altered for local use.

Combining the two would mean amending the locations the various bash command are running in, and I'd see that as an optimization too far at this stage. It would also complicate the creation of a local use _"Hacker's"_ `Dockerfile`

*If* there are two layers, the clean-up statements need to be repeated as you can only clean up within the current layer.

File: docker/Dockerfile (hunk 3 above)
Comment (jason-fox) [reply docker/Dockerfile]: Fixed e290fb5

File: docker/README.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+It is based on the
+[IoT Agent Node.js Library](https://github.com/telefonicaid/iotagent-node-lib).
+Further general information about the FIWARE IoT Agents framework, its
+architecture and the common interaction model can be found in the library's
+GitHub repository.
+
+This project is part of [FIWARE](https://www.fiware.org/). For more information
+check the FIWARE Catalogue entry for the
+[IoT Agents](https://github.com/Fiware/catalogue/tree/master/iot-agents).
+
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply docker/README.md]: Yes, I'll add an explanatory text of the "How to Use Dockerfile" text into the `README`. I'd prefer not to remove the comment from the `Dockerfile` itself though since a **developer** may not think to look at the `README` 

I know I would tend to look direct at the `Dockerfile` myself. It's nice to find instructions within the one file you are editing.

File: docker/README.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+## Configuration with environment variables
+Some of the more common variables can be configured using environment variables. The ones overriding general parameters
+in the `config.iota` set are described in the [IoTA Library Configuration manual](https://github.com/telefonicaid/iotagent-node-lib#configuration).
+
+The ones specifically relating to Ultralight bindings are described in the following table.
```
[the remark this answers is not in the document]
Comment (jason-fox) [reply docker/README.md]: Link added. Even if the table goes stale, the explanatory text directs you onwards for the latest information. Fixed fa9481c

[the remark this answers is not in the document]
Comment (fgalan) [reply docker/README.md]: Putting myself in developers's shoes, I prefer to navigate a link to get the accurate list of settings (please, don't assume our target developers are so lazy... they use to be smart people ;) that spending hours to find a problem due to the env var which name I though was correct actually it wasn't.

In fact, a link is already added for library env vars. Why don't do the same for agent-specific env vars?

With regards to:

> Most good Docker Hub README files list the ENV variables you can use, we should attempt to follow good practice.

Do these projects duplicate env vars information (in Docker hub and upstream software documentation)? Or they only have the copy of the env vars in the Docker hub?

I think it would be a good idea to gets more opinions on this. @dcalvoalonso @AlvaroVega what do you think?

File: docker/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ARG DOWNLOAD_TYPE=latest
+
+ENV GITHUB_ACCOUNT=${GITHUB_ACCOUNT}
+ENV GITHUB_REPOSITORY=${GITHUB_REPOSITORY}
+ENV DOWNLOAD_TYPE=${DOWNLOAD_TYPE}
+
+MAINTAINER FIWARE IoTAgent Team. Telefónica I+D
+
+#
+# The following line retrieves the latest source code from GitHub.
```
Comment (fgalan) [inline docker/Dockerfile]: A couple of suggestions in the text to make it more clear:

The following line... -> The following RUN statement

Alternatively for local development, just copy this Dockerfile into file the root of the repository and copy over your local source using : -> Alternatively for local development, just copy this Dockerfile into file the root of the repository and replace the whole RUN statement by the following COPY statement in your local source using :

File: docker/README.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+* By default, the `Dockerfile` retrieves the **latest** version of the codebase direct from GitHub (the `build-arg` is optional):
+
+```console
+docker build -t iot-agent . --build-arg DOWNLOAD_TYPE=latest
+```
+
+* You can alter this to obtain the last **stable** release run this `Dockerfile` with the build argument `DOWNLOAD_TYPE=stable`
+
+```console
+docker build -t iot-agent . --build-arg DOWNLOAD_TYPE=stable
```
Comment (fgalan) [inline docker/README.md]: This is fine to build the latest stable tagged release but, what about to build a particular tag, i.e. something like:

```
 docker build -t iot-agent . --build-arg DOWNLOAD_TYPE=1.6.0
```

By the way the current `if [ ${DOWNLOAD_TYPE}` works in Dockerfile I think currently it is not possible, but it would be great to modify so we can also have that possibility.

File: docker/README.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    hostname: mongo-db
+    container_name: db-mongo
+    ports:
+        - "27017:27017"
+    command: --bind_ip_all --smallfiles
+    volumes:
+       - mongodb:/data
+
+  orion:
+    image: fiware/orion:2.0.0
```
Comment (fgalan) [inline docker/README.md]: We have in a line above:

```
 image: fiware/iotagent-ul
```

so I'd suggest to use here:

```
    image: fiware/orion:2.0.0
```

File: docker/README.md [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+  mongodb:
+    image: mongo:3.6
+    hostname: mongo-db
+    container_name: db-mongo
+    ports:
+        - "27017:27017"
+    command: --bind_ip_all --smallfiles
+    volumes:
+       - mongodb:/data
```
Comment (fgalan) [inline docker/README.md]: A comment should be done about the need of a /data directory in the hosting machine so this example may work. The last sentence could be modified in the following way to cope with it:

```
If the `IOTA_REGISTRY_TYPE=mongodb`, a [MongoDB](https://www.mongodb.com/) database instance is also required. 
The example assumes that you have a `/data` directory in your hosting system in order to hold database files (please 
change if you want to use another one in your case)"
```

File: docker/README.md [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    depends_on:
+        - mongodb
+    expose:
+        - "1026"
+    ports:
+        - "1026:1026"
+    command: -dbhost mongodb
+```
+
+
```
Comment (fgalan) [inline docker/README.md]: Unneed extra blank lines?

File: docker/README.md (hunk 4 above)
Comment (AlvaroVega) [reply docker/README.md]: +1 to @fgalan, just a link should be enough

File: docker/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlvaroVega) [reply docker/README.md]: +1 to @fgalan, a link should be enough

File: docker/Dockerfile (hunk 1 above)
Comment (dcalvoalonso) [reply docker/Dockerfile]: +1 to https://github.com/telefonicaid/iotagent-ul/pull/310/files#r236232373

File: docker/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dcalvoalonso) [reply docker/README.md]: +1 to @fgalan, repeated information is more difficult to maintain and at the end, something is not updated and causes more problems than a single link

[the remark this answers is not in the document]
Comment (jason-fox) [reply docker/README.md]: Documentation usability is negatively affected if you need to click through to a greater depth, assuming someone looks up Docker information on Docker Hub (which is a fair assumption) and **fails to find it there**, effectively you are saying:

## [Thank you Mario, but our princess is in another castle](https://knowyourmeme.com/memes/but-our-princess-is-in-another-castle)

Which is mildly annoying. 😠  - Why do we want to annoy our developers?

> Do these projects duplicate env vars information (in Docker hub and upstream software documentation)? Or they only have the copy of the env vars in the Docker hub?

As an example of Good practice, look at the following for **MySQL**:

*  The real documentation - https://dev.mysql.com/doc/refman/8.0/en/docker-mysql-more-topics.html#docker-environment-variables
*  The equivalent on Docker Hub:  https://hub.docker.com/_/mysql/

This to me looks like a reasonable compromise on this would be to include an opening paragraph with basic docker usage,  a prominent link in it and at least the major configuration elements in it. I'm making the assumption here that the major `ENV` variables are unlikely to be renamed.

File: docker/Dockerfile (hunk 1 above)
Comment (jason-fox) [reply docker/Dockerfile]: Fixed - Explanation added 03e74ec

File: docker/Dockerfile (hunk 7 above)
Comment (jason-fox) [reply docker/Dockerfile]: Fixed - text added -  03e74ec

File: docker/README.md (hunk 8 above)
Comment (jason-fox) [reply docker/README.md]: Bash script expanded - Comments added
Fixed - 03e74ec

File: docker/README.md (hunk 10 above)
Comment (jason-fox) [reply docker/README.md]: Fixed - 13e42ee

File: docker/README.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+check the FIWARE Catalogue entry for the
+[IoT Agents](https://github.com/Fiware/catalogue/tree/master/iot-agents).
+
+## How to use this image
+
+The IoT Agent must be instantiated and connected to an instance of the [Orion Context Broker](https://fiware-orion.readthedocs.io/en/latest/), a sample `docker-compose` file can be found below.
+
+If the `IOTA_REGISTRY_TYPE=mongodb`, a [MongoDB](https://www.mongodb.com/) database instance is also required - the example below assumes that you
+have a `/data` directory in your hosting system in order to hold database
+files (please amend the attached volume to suit your own configuration)"
```
Comment (fgalan) [inline docker/README.md]: Typo (I guess): `"` -> `.`

File: docker/Dockerfile (hunk 2 above)
Comment (fgalan) [reply docker/Dockerfile]: 4% is fine. Thanks for the info.

NTC

File: docker/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (fgalan) [reply docker/README.md]: If we restrict the list of env vars to be described to the ones in the docker-compose.yml used as example fine with me. In fact, this will be a way of documenting the docker-compose.yml embedded in the README.md. That is:

```
        - "IOTA_CB_HOST=orion"
        - "IOTA_CB_PORT=1026"
        - "IOTA_NORTH_PORT=4041"
        - "IOTA_REGISTRY_TYPE=mongodb"
        - "IOTA_MONGO_HOST=mongo-db"
        - "IOTA_MONGO_PORT=27017"
        - "IOTA_MONGO_DB=iotagent-ul"
        - "IOTA_HTTP_PORT=7896"
        - "IOTA_PROVIDER_URL=http://iot-agent:4041"
```

In addition, I'd suggest not using the current table style (which maps env var with config.js element) but a short description for each one, in the same fashion that the https://hub.docker.com/_/mysql/ you cite.

If others develpers are ok with this (@dcalvoalonso @AlvaroVega what do you think?) please go ahead.

[the remark this answers is not in the document]
Comment (AlvaroVega) [reply docker/README.md]: If only some of the env vars are going to be in that page I'd prefer not use a table, just a short description.

[the remark this answers is not in the document]
Comment (jason-fox) [reply docker/README.md]: @fgalan @AlvaroVega  - a limited bullet list sounds like a good idea - a list appears to be less **comprehensive** than a formal table, but still provides sufficient documentation to be able to run a simple dockerized solution.

Implicitly it provides the following information:

1.  The system can be configured using Environment variables
2.  Some major environment variables are documented here on **DockerHub** - we are able to write a good README 
3.  Environment variables are documented in more detail **elsewhere** (avoid duplicate effort)
4.  You won't find **everything** here, but we are considerate enough to get you started and provide a link.

In summary this approach states -  _we aim to be nice to the developer._

[the remark this answers is not in the document]
Comment (fgalan) [reply docker/README.md]: @dcalvoalonso what dou you think?

[the remark this answers is not in the document]
Comment (jason-fox) [reply docker/README.md]: Duplicate Table replaced with shortened list and sign-posted links to the actual documentation.

Fixed - 92fe3e19

File: docker/README.md (hunk 12 above)
Comment (jason-fox) [reply docker/README.md]: Fixed - 2120915c

File: docker/README.md (hunk 4 above)
Comment (fgalan) [reply docker/README.md]: This comment is yet pending, but given that PR #301 is not yet mature is this is probably about to merge, I'd suggest to move the comment to PR #301 so the link to docker/README.md be included there.

NTC

**Comment by fgalan** [comment]
(Travis can be ignored in this PR. Travis is failing probably due to the PR was branched in a moment in which master wasn't travis-ok. It doesn't matter, as this PR doesn't touch IOTA code)

**Comment by dcalvoalonso** [comment]
Great job @jason-fox and amazing discussion, I have learnt a lot reading your comments!

Sorry for not having been more proactive but these are being complicated weeks...

Of course, LGTM.

**Comment by fgalan** [comment]
I have changed the build setting in dockerhub at https://hub.docker.com/r/fiware/iotagent-ul/ to use `/docker` instead of `/` but it seems the old README.md is still rendered... Maybe some additional setting needs to be done?

**Comment by fgalan** [comment]
Now it renders correctly

Weird... maybe it was a timming issue. Dockerhub may take some undefined time (several minutes) to update the rendered document.
